src/esp32/src/flash.c

```diff
@@ -4,16 +4,148 @@
+}
+
+// TODO: common code with esp32_flash_spi_wait_ready
+static uint32_t esp32_flash_exec_usr_cmd(uint32_t cmd)
+{
+    uint32_t status_value = ESP_ROM_SPIFLASH_BUSY_FLAG;
+
+    while (ESP_ROM_SPIFLASH_BUSY_FLAG == (status_value & ESP_ROM_SPIFLASH_BUSY_FLAG)) {
+        WRITE_PERI_REG(PERIPHS_SPI_FLASH_STATUS, 0); /* clear register */
+        WRITE_PERI_REG(PERIPHS_SPI_FLASH_CMD, SPI_USR | cmd); //TODO cmd is bit(0), this is a useless bit
+        while (READ_PERI_REG(PERIPHS_SPI_FLASH_CMD) != 0)
+            ;
+        status_value = READ_PERI_REG(PERIPHS_SPI_FLASH_STATUS) & g_rom_flashchip.status_mask;
+    }
+
+    return status_value;
+}
+
+static void esp32_flash_spi_wait_ready(void)
+{
+    uint32_t status_value = ESP_ROM_SPIFLASH_BUSY_FLAG;
+
+    while (ESP_ROM_SPIFLASH_BUSY_FLAG == (status_value & ESP_ROM_SPIFLASH_BUSY_FLAG)) {
+        WRITE_PERI_REG(PERIPHS_SPI_FLASH_STATUS, 0); /* clear register */
+        WRITE_PERI_REG(PERIPHS_SPI_FLASH_CMD, SPI_FLASH_RDSR);
+        while (READ_PERI_REG(PERIPHS_SPI_FLASH_CMD) != 0)
+            ;
+        status_value = READ_PERI_REG(PERIPHS_SPI_FLASH_STATUS) & g_rom_flashchip.status_mask;
+    }
+}
+
+static uint32_t esp32_flash_spi_cmd_run(uint32_t cmd,
+                                        uint8_t data_bits[],
+                                        uint32_t data_bits_num,
+                                        uint32_t read_bits_num)
+{
+    uint32_t old_spi_usr = READ_PERI_REG(PERIPHS_SPI_FLASH_USRREG);
+    uint32_t old_spi_usr2 = READ_PERI_REG(PERIPHS_SPI_FLASH_USRREG2);
+    uint32_t flags = SPI_USR_COMMAND;
+
+    esp32_flash_spi_wait_ready();
+
+    if (read_bits_num > 0) {
+        flags |= SPI_USR_MISO;
+        WRITE_PERI_REG(PERIPHS_SPI_MISO_DLEN_REG, read_bits_num - 1);
+    }
+    if (data_bits_num > 0) {
+        flags |= SPI_USR_MOSI;
+        WRITE_PERI_REG(PERIPHS_SPI_MOSI_DLEN_REG, data_bits_num - 1);
+    }
+
+    WRITE_PERI_REG(PERIPHS_SPI_FLASH_USRREG, flags);
+    WRITE_PERI_REG(PERIPHS_SPI_FLASH_USRREG2, (7 << SPI_USR2_DLEN_SHIFT) | cmd);
+    if (data_bits_num == 0) {
+        WRITE_PERI_REG(PERIPHS_SPI_FLASH_C0, 0);
+    } else {
+        for (uint32_t i = 0; i <= data_bits_num / 32; i += 32) {
+            WRITE_PERI_REG(PERIPHS_SPI_FLASH_C0 + i / 8, *((uint32_t *)&data_bits[i / 8]));
+        }
+    }
+    esp32_flash_exec_usr_cmd(0);
+    uint32_t status = READ_PERI_REG(PERIPHS_SPI_FLASH_C0);
+    /* restore some SPI controller registers */
+    WRITE_PERI_REG(PERIPHS_SPI_FLASH_USRREG, old_spi_usr);
+    WRITE_PERI_REG(PERIPHS_SPI_FLASH_USRREG2, old_spi_usr2);
+
+    return status;
+}
+
+static inline uint32_t device_id_from_spi_rdid()
+{
+
+    uint32_t rdid = esp32_flash_spi_cmd_run(SPI_FLASH_CMD_RDID, NULL, 0, 24);
+    return ((rdid & 0xff) << 16) | (rdid & 0xff00) | ((rdid & 0xff0000) >> 16);;
+}
+
+uint32_t stub_target_flash_device_id(void)
+{
+    STUB_LOG_TRACEF("Uninit g_rom_flashchip.device_id: 0x%x\n", g_rom_flashchip.device_id);
```

Comment:
General suggestion; trace logs are Ok but if they are much, they will not be useful.  For example; we can print device_id after return in the caller function

Reply:
Yes, I do. Only for the current stage of work it is useful for me to check different chips, later then it will be cleaned up. Not sure that we need to add another macro, maybe put a comment only here